include/Signal.hpp

```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <algorithm>
+#include <mutex>
+#include <memory>
+#include <vector>
+#include <functional>
+
+namespace Framework
+{
+
+	template<typename> class CSignal;
+	template<typename T, typename... Args>
+	class CSignal<T (Args...)>
+	{
+
+	public:
+		class CConnection;
+
+		typedef std::shared_ptr<CConnection> CConnectionPtr;
+		typedef std::weak_ptr<CConnection> CConnectionWeakPtr;
+		typedef std::vector<CConnectionPtr> CConnectionListPtr;
+
+		typedef CSignal<T (Args...)> CSignalProto;
+		typedef std::function<void()> CConectionFunction;
+		typedef std::function<T (Args...)> CSlotFunction;
+		typedef std::weak_ptr<const CSlotFunction> CSlotFunctionPtr;
+
+		CSignal() = default;
+
+		CConnectionPtr Connect(const CSlotFunction& func, bool oneShot = false)
+		{
+			if(func)
+			{
+				std::unique_lock<std::mutex> lock(m_lock);
+
+				CConnectionPtr connection = std::make_shared<CConnection>(func, oneShot);
+				m_connections.push_back(connection);
+
+				return connection;
+			}
+
+			return nullptr;
+		}
+
+		CConnectionPtr ConnectOnce(const CSlotFunction& func)
+		{
+			return Connect(func, true);
+		}
+
+		void Reset()
+		{
+			std::unique_lock<std::mutex> lock(m_lock);
+
+			m_connections.clear();
+		}
+
+		void operator()(Args... args)
+		{
+			std::unique_lock<std::mutex> lock(m_lock);
+
+			m_connections.erase(
+				std::remove_if(
+					m_connections.begin(), 
+					m_connections.end(),
+					[&](CConnectionWeakPtr& connection)
+					{
+						auto connectionPtr = connection.lock();
+						if(connectionPtr)
+							(*connectionPtr)(args...);
+						return !connectionPtr || (*connectionPtr).IsOneShot();
+					}
+				), 
+				m_connections.end()
+			);
+		}
+
+		class CConnection
+		{
+			public:
+				CConnection(const CSlotFunction slot, bool oneShot)
+				: m_slot(slot)
+				, m_oneShot(oneShot)
+				{
+				}
+
+				~CConnection()
+				{
+				}
+
+				void operator()(Args... args)
+				{
+					if(m_slot)
```

Comment:
We don't need to check if `m_slot` is empty here. It'll always be something (and it will assert in `Connect` if someone tries to pass something empty).